### PR TITLE
refactor(ivy): remove ngBaseDef

### PR DESCRIPTION
**Pull request by crisbeto**
Removes `ngBaseDef` from the compiler and any runtime code that was still referring to it. In the cases where we'd previously generate a base def we now generate a definition for an abstract directive.

### CODE REVIEW

File: packages/compiler-cli/src/ngtsc/annotations/src/directive.ts [hunk 1]
```diff
@@ -233,21 +218,80 @@ export function extractDirectiveMetadata(
...
+  const members = reflector.getMembersOfClass(clazz);
+
+  // Precompute a list of ts.ClassElements that have decorators. This includes things like @Input,
+  // @Output, @HostBinding, etc.
+  const decoratedElements =
+      members.filter(member => !member.isStatic && member.decorators !== null);
+
+  const coreModule = isCore ? undefined : '@angular/core';
+
+  // Construct the map of inputs both from the decorated fields.
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngtsc/annotations/src/directive.ts]: This sentence doesn't make sense; and probably isn't really needed.

File: packages/compiler-cli/src/ngtsc/annotations/src/directive.ts [hunk 2]
```diff
@@ -233,21 +218,80 @@ export function extractDirectiveMetadata(
...
+  const inputs = parseDecoratedFields(
+      filterToMembersWithDecorator(decoratedElements, 'Input', coreModule), evaluator,
+      resolveInput);
+
+  // And outputs.
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngtsc/annotations/src/directive.ts]: Superfluous comment.

File: packages/compiler-cli/src/ngtsc/annotations/src/directive.ts [hunk 3]
```diff
@@ -233,21 +218,80 @@ export function extractDirectiveMetadata(
...
+  const outputs = parseDecoratedFields(
+      filterToMembersWithDecorator(decoratedElements, 'Output', coreModule), evaluator,
+      resolveOutput) as{[field: string]: string};
+  // Construct the list of queries.
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngtsc/annotations/src/directive.ts]: Should be an empty line before this comment to keep consistent with the other formatting in this function.

File: packages/compiler-cli/src/ngtsc/annotations/src/directive.ts [hunk 4]
```diff
@@ -587,6 +592,56 @@ export function extractHostBindings(
...
+/** Extracts host bindings from an Angular class decorator (e.g. Directive or Component). */
+function extractHostBindingsFromDecoratorMetadata(
+    evaluator: PartialEvaluator, metadata: Map<string, ts.Expression>): ParsedHostBindings {
+  let hostMetadata: StringMap<string|Expression> = {};
+  if (metadata.has('host')) {
+    const expr = metadata.get('host') !;
+    const hostMetaMap = evaluator.evaluate(expr);
+    if (!(hostMetaMap instanceof Map)) {
+      throw new FatalDiagnosticError(
+          ErrorCode.DECORATOR_ARG_NOT_LITERAL, expr, `Decorator host metadata must be an object`);
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngtsc/annotations/src/directive.ts]:
```suggestion
          ErrorCode.DECORATOR_ARG_NOT_LITERAL, expr, `Decorator host metadata must evaluate to an object`);
```

**Review by petebacondarwin — APPROVED** [review]

**Comment by AndrewKushnir** [comment]
Initial set of presubmits:
- [VE Presubmit](https://test.corp.google.com/ui#id=OCL:275944252:BASE:275944457:1571698954529:3cffe091)
- [Ivy Presubmit](https://test.corp.google.com/ui#id=OCL:275944252:BASE:275948176:1571700081146:7f253a98)

**Comment by AndrewKushnir** [comment]
Quick update after running presubmit in g3: it looks like there are some regressions, but they seem to be caused by the fact that there is a problem with inheriting inputs from base classes (annotated with `@Directive()`), which is used in this PR as a replacement for `ngBaseDef`. The problem should be resolved by PR #33131, but it's hard to tell whether it's the only problem in g3 or not. My proposal is to land PR #33131 first, rebase this PR and run g3 presubmit again (if time permits). As an option, we can create a draft PR that would include the changes from this PR and PR #33131 to see if there are any regressions in g3. Thank you.

**Comment by crisbeto** [comment]
Are those compile-time errors or runtime? If I remember correctly, we have a lot of inheritance tests for the runtime which are passing.

**Comment by AndrewKushnir** [comment]
@crisbeto yes, the errors happen at runtime, so I guess I was incorrect assuming that #33131 will fix those. Let me collect more information and get back with specific use-cases that regressed. Thank you.

File: packages/compiler-cli/src/ngtsc/annotations/src/directive.ts [hunk 5]
```diff
@@ -19,42 +19,79 @@ import {AnalysisOutput, CompileResult, DecoratorHandler, DetectResult, HandlerPr
...
-      return undefined;
-    }
-    const decorator = findAngularDecorator(decorators, 'Directive', this.isCore);
-    if (decorator !== undefined) {
-      return {
-        trigger: decorator.node,
-        metadata: decorator,
-      };
+      const angularField = this.reflector.getMembersOfClass(node).find(member => {
+        if (isNgOnChanges(member)) {
```
Comment (kara) [inline packages/compiler-cli/src/ngtsc/annotations/src/directive.ts]: Let's make this all lifecycle hooks. In version 10, we plan on asking users to add a decorator to any class with an Angular lifecycle hook, so we should be consistent here.

File: packages/compiler-cli/src/ngtsc/annotations/src/directive.ts [hunk 6]
```diff
@@ -19,42 +19,79 @@ import {AnalysisOutput, CompileResult, DecoratorHandler, DetectResult, HandlerPr
...
       private reflector: ReflectionHost, private evaluator: PartialEvaluator,
       private metaRegistry: MetadataRegistry, private defaultImportRecorder: DefaultImportRecorder,
       private isCore: boolean) {}
 
   readonly precedence = HandlerPrecedence.PRIMARY;
 
-  detect(node: ClassDeclaration, decorators: Decorator[]|null): DetectResult<Decorator>|undefined {
+  detect(node: ClassDeclaration, decorators: Decorator[]|null):
+      DetectResult<Decorator|null>|undefined {
+    // If the class is undecorated, check if any of the fields have Angular decorators,
```
Comment (kara) [inline packages/compiler-cli/src/ngtsc/annotations/src/directive.ts]:
```suggestion
    // If the class is undecorated, check if any of the fields have Angular decorators or lifecycle hooks,
```

File: packages/core/src/render3/jit/directive.ts [hunk 7]
```diff
@@ -336,10 +280,40 @@ function isViewQuery(value: any): value is Query {
...
   return value.split(',').map(piece => piece.trim());
 }
+
+function usesNgOnChanges(type: Type<any>): boolean {
+  const prototype = type.prototype;
+  return prototype && prototype.hasOwnProperty('ngOnChanges');
+}
+
+function shouldAddAbstractDirective(type: Type<any>): boolean {
+  if (usesNgOnChanges(type)) {
```
Comment (kara) [inline packages/core/src/render3/jit/directive.ts]: Same here. Shouldn't it be all hooks?

**Review by kara — CHANGES_REQUESTED** [review]

**Comment by crisbeto** [comment]
The feedback has been addressed.

File: packages/compiler-cli/src/ngtsc/annotations/src/directive.ts [hunk 8]
```diff
@@ -19,42 +19,88 @@ import {AnalysisOutput, CompileResult, DecoratorHandler, DetectResult, HandlerPr
...
+    // hooks, and if they do, label the class as an abstract directive.
+    if (!decorators) {
+      const angularField = this.reflector.getMembersOfClass(node).find(member => {
+        if (isLifecycleHook(member, 'ngOnChanges') || isLifecycleHook(member, 'ngOnInit') ||
+            isLifecycleHook(member, 'ngOnDestroy') || isLifecycleHook(member, 'ngDoCheck') ||
+            isLifecycleHook(member, 'ngAfterViewInit') ||
+            isLifecycleHook(member, 'ngAfterViewChecked') ||
+            isLifecycleHook(member, 'ngAfterContentInit') ||
+            isLifecycleHook(member, 'ngAfterContentChecked')) {
+          return true;
```
Comment (kara) [inline packages/compiler-cli/src/ngtsc/annotations/src/directive.ts]: Move hooks into an array like we did for FIELD_DECORATORS?

```typescript
if (LIFECYCLE_HOOKS.some(hook => isLifecycleHook(member, hook)) {
```

**Review by kara — APPROVED** [review]
LGTM, with one stylistic nit

**Comment by crisbeto** [comment]
Addressed the last comment.

**Review by kara — APPROVED** [review]
LGTM

File: packages/compiler-cli/src/ngtsc/annotations/src/directive.ts [hunk 9]
```diff
@@ -19,42 +19,87 @@ import {AnalysisOutput, CompileResult, DecoratorHandler, DetectResult, HandlerPr
...
-    if (decorator !== undefined) {
-      return {
-        trigger: decorator.node,
-        metadata: decorator,
-      };
+    // If the class is undecorated, check if any of the fields have Angular decorators or lifecycle
+    // hooks, and if they do, label the class as an abstract directive.
+    if (!decorators) {
+      const angularField = this.reflector.getMembersOfClass(node).find(member => {
+        if (LIFECYCLE_HOOKS.some(hookName => isLifecycleHook(member, hookName))) {
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/annotations/src/directive.ts]:
```suggestion
        if (!member.isStatic && member.kind === ClassMemberKind.Method && LIFECYCLE_HOOKS.has(member.name)) {
```

File: packages/compiler-cli/src/ngtsc/annotations/src/directive.ts [hunk 10]
```diff
@@ -19,42 +19,87 @@ import {AnalysisOutput, CompileResult, DecoratorHandler, DetectResult, HandlerPr
...
   }
 
-  analyze(node: ClassDeclaration, decorator: Decorator): AnalysisOutput<DirectiveHandlerData> {
-    const directiveResult = extractDirectiveMetadata(
-        node, decorator, this.reflector, this.evaluator, this.defaultImportRecorder, this.isCore);
-    const analysis = directiveResult && directiveResult.metadata;
+  analyze(node: ClassDeclaration, decorator: Decorator|null): AnalysisOutput<DirectiveHandlerData> {
+    let analysis: R3DirectiveMetadata|undefined;
+
+    if (decorator) {
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/annotations/src/directive.ts]:
```suggestion
    if (decorator !== null) {
```

File: packages/compiler-cli/src/ngtsc/annotations/src/directive.ts [hunk 11]
```diff
@@ -108,13 +153,12 @@ export class DirectiveDecoratorHandler implements
  * appear in the declarations of an `NgModule` and additional verification is done when processing
  * the module.
  */
-export function extractDirectiveMetadata(
+export function extractDirectiveMetadataFromDecoratedClass(
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/annotations/src/directive.ts]: Don't refactor this method. Instead, change `decorator` to take `Decorator|null`.

File: packages/compiler-cli/src/ngtsc/annotations/src/directive.ts [hunk 12]
```diff
@@ -580,9 +607,64 @@ export function extractHostBindings(
           bindings.listeners[eventName] = `${member.name}(${args.join(',')})`;
         });
       });
+
+  return bindings;
+}
+
+/** Extracts host bindings from an Angular class decorator (e.g. Directive or Component). */
+function extractHostBindingsFromDecoratorMetadata(
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/annotations/src/directive.ts]: Same here. No need for such a large refactoring - the original `extractHostBindings` should be fine.

File: packages/compiler/src/jit_compiler_facade.ts [hunk 13]
```diff
@@ -168,24 +168,6 @@ export class CompilerFacadeImpl implements CompilerFacade {
         factoryRes.factory, angularCoreEnv, sourceMapUrl, factoryRes.statements);
   }
 
-  compileBase(angularCoreEnv: CoreEnvironment, sourceMapUrl: string, facade: R3BaseMetadataFacade):
```
Comment (alxhub) [inline packages/compiler/src/jit_compiler_facade.ts]: 🍾

File: packages/compiler/src/render3/view/api.ts [hunk 14]
```diff
@@ -13,41 +13,11 @@ import {ParseSourceSpan} from '../../parse_util';
 import * as t from '../r3_ast';
 import {R3DependencyMetadata} from '../r3_factory';
 
-
 /**
- * Information needed to compile a directive for the render3 runtime.
+ * Information needed to compile a directive for the render3 runtime, extracted only from the class
+ * members, excluding any Angular decorator metadata.
  */
-export interface R3DirectiveMetadata {
```
Comment (alxhub) [inline packages/compiler/src/render3/view/api.ts]: No, leave this as is.

**Comment by AndrewKushnir** [comment]
- [VE Presubmit](http://test/OCL:276583247:BASE:276583294:1571957470486:784535e8)
- [Ivy Presubmit](http://test/OCL:276583247:BASE:276583452:1571957521894:1afe8ca9)

**Review by alxhub — CHANGES_REQUESTED** [review]

**Comment by crisbeto** [comment]
The latest set of feedback has been addressed.

File: packages/compiler-cli/src/ngtsc/annotations/src/directive.ts (hunk 12 above)
Comment (crisbeto) [reply packages/compiler-cli/src/ngtsc/annotations/src/directive.ts]: Yeah those are unused now I just forgot to delete them the first time I pushed. They're gone now.

File: packages/compiler-cli/src/ngtsc/annotations/src/directive.ts [hunk 15]
```diff
@@ -19,42 +19,64 @@ import {AnalysisOutput, CompileResult, DecoratorHandler, DetectResult, HandlerPr
...
 import {generateSetClassMetadataCall} from './metadata';
-import {findAngularDecorator, getValidConstructorDependencies, readBaseClass, unwrapExpression, unwrapForwardRef} from './util';
+import {findAngularDecorator, getValidConstructorDependencies, isAngularDecorator, readBaseClass, unwrapExpression, unwrapForwardRef} from './util';
 
 const EMPTY_OBJECT: {[key: string]: string} = {};
+const FIELD_DECORATORS = [
+  'Input', 'Output', 'ViewChild', 'ViewChildren', 'ContentChild', 'ContentChildren', 'HostBinding',
+  'HostListener'
+];
+const LIFECYCLE_HOOKS = [
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/annotations/src/directive.ts]: Wrap in a `Set` for faster searching.

File: packages/compiler-cli/src/ngtsc/annotations/src/directive.ts [hunk 16]
```diff
@@ -109,15 +131,14 @@ export class DirectiveDecoratorHandler implements
  * the module.
  */
 export function extractDirectiveMetadata(
-    clazz: ClassDeclaration, decorator: Decorator, reflector: ReflectionHost,
+    clazz: ClassDeclaration, decorator: Decorator | null, reflector: ReflectionHost,
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/annotations/src/directive.ts]:
```suggestion
    clazz: ClassDeclaration, decorator: Decorator|null, reflector: ReflectionHost,
```

Surprised clang-format doesn't do this.

File: packages/compiler/src/render3/view/api.ts (hunk 14 above)
Comment (crisbeto) [reply packages/compiler/src/render3/view/api.ts]: Done.

File: packages/compiler/src/render3/view/api.ts [hunk 17]
```diff
@@ -85,11 +54,41 @@ export interface R3DirectiveMetadata {
    */
   outputs: {[field: string]: string};
 
+  /**
```
Comment (alxhub) [inline packages/compiler/src/render3/view/api.ts]: Undo this rearrangement :)

**Review by alxhub — APPROVED** [review]

**Comment by crisbeto** [comment]
Addressed more of the feedback.

File: packages/compiler-cli/src/ngtsc/annotations/src/directive.ts (hunk 16 above)
Comment (crisbeto) [reply packages/compiler-cli/src/ngtsc/annotations/src/directive.ts]: Actually it seems to reformat it to the one with spaces when I remove the space. I'll leave it as is.

**Comment by AndrewKushnir** [comment]
@crisbeto FYI, there are some failures in CI after the latest changes...

**Comment by crisbeto** [comment]
Blocked on https://github.com/angular/angular/pull/32987.

**Comment by AndrewKushnir** [comment]
FYI, VE and Ivy presubmits were successful (g3 regression was resolved) at the point when I started them ([see my comment here](https://github.com/angular/angular/pull/33264#issuecomment-546133821)). I believe CI was "green" at that time as well, so I'm wondering if we can roll back recent changes in this PR (based on CR) if they are not critical, so we can unblock this PR. The remaining comments/improvements (if they are not critical) can be performed as a followup, especially given that we became blocked on #32987.

**Comment by alxhub** [comment]
Caretaker: these are being done preemptively. Once #32987 lands, this should be in a good state.

[Presubmit](http://test/OCL:276735680:BASE:276735745:1572030696588:240079d5)
[Ivy Presubmit](http://test/OCL:276735680:BASE:276735873:1572030752858:391515af)

**Pull request closed without merging**